File: sdk/typescript/src/auth.test.ts
import { test, expect } from "vitest";
import { resolveAuth } from "./auth.ts";

test("undefined auth returns empty headers", async () => {
  const resolve = resolveAuth(undefined);
  const headers = await resolve();
  expect(headers).toEqual({});
});

test("string auth returns Bearer token", async () => {
  const resolve = resolveAuth("sk-abc123");
  const headers = await resolve();
  expect(headers).toEqual({ Authorization: "Bearer sk-abc123" });
});

test("object auth returns headers as-is", async () => {
  const resolve = resolveAuth({ "X-API-Key": "abc", "X-Workspace": "ws-1" });
  const headers = await resolve();
  expect(headers).toEqual({ "X-API-Key": "abc", "X-Workspace": "ws-1" });
});

test("function auth calls provider", async () => {
  let callCount = 0;
  const resolve = resolveAuth(async () => {
    callCount++;
    return { Authorization: `Bearer token-${callCount}` };
  });

  const h1 = await resolve();
  expect(h1).toEqual({ Authorization: "Bearer token-1" });

  const h2 = await resolve();
  expect(h2).toEqual({ Authorization: "Bearer token-2" });
  expect(callCount).toBe(2);
});

File: sdk/go/auth_test.go
package sandbox

import (
	"context"
	"testing"
)

func TestStringAuth(t *testing.T) {
	auth := &StringAuth{Token: "sk-abc123"}
	headers, err := auth.Resolve(context.Background())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	want := "Bearer sk-abc123"
	if headers["Authorization"] != want {
		t.Fatalf("got %q, want %q", headers["Authorization"], want)
	}
}

func TestHeadersAuth(t *testing.T) {
	auth := &HeadersAuth{Headers: map[string]string{
		"X-API-Key": "key123",
		"X-Tenant":  "acme",
	}}
	headers, err := auth.Resolve(context.Background())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if headers["X-API-Key"] != "key123" {
		t.Fatalf("got %q, want %q", headers["X-API-Key"], "key123")
	}
	if headers["X-Tenant"] != "acme" {
		t.Fatalf("got %q, want %q", headers["X-Tenant"], "acme")
	}
	// Verify it returns a copy, not the original map.
	headers["X-API-Key"] = "modified"
	h2, _ := auth.Resolve(context.Background())
	if h2["X-API-Key"] != "key123" {
		t.Fatal("HeadersAuth should return a copy of headers")
	}
}

func TestProviderAuth(t *testing.T) {
	called := false
	auth := &ProviderAuth{
		Provider: func(ctx context.Context) (map[string]string, error) {
			called = true
			return map[string]string{"Authorization": "Bearer dynamic-token"}, nil
		},
	}
	headers, err := auth.Resolve(context.Background())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !called {
		t.Fatal("provider was not called")
	}
	if headers["Authorization"] != "Bearer dynamic-token" {
		t.Fatalf("got %q, want %q", headers["Authorization"], "Bearer dynamic-token")
	}
}

func TestProviderAuthNilProvider(t *testing.T) {
	auth := &ProviderAuth{}
	headers, err := auth.Resolve(context.Background())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(headers) != 0 {
		t.Fatalf("expected empty headers, got %v", headers)
	}
}
